### PR TITLE
Contenteditable table headers

### LINKED ISSUE #14113: Add support for reporting row and column headers in contenteditable HTML elements.
### Steps to reproduce:
**Repro Steps:**
1. Open the URL: https://ckeditor.com/ckeditor-4/demo/
1. Turn on NVDA.
2. Turn on NVDA Focus mode.
3. Add a new table and add column headers for the table
4. Now using ctrl + alt + arrow keys to navigate from one cell to another
5. Verify whether NVDA is announcing column headers or not.

Code pen for minimal sample: https://codepen.io/testaccount01/pen/rNvxdwJ

### Actual behavior (Speech Viewer):-
For 1st table: -
row 2
column 1
row 1  a
row 2
row 3
column 2
column 2  b
column 1
row 3
column 2

For 2nd Table (Speech Viewer) :-
column 2  H2
column 1  H1
row 2  R1
column 2
column 3
row 1  H3
row 2
row 3
column 2
row 1  H3

NVDA is not announcing row or column headers.
-->

### Expected behavior:
NVDA should announce row or column headers when we navigate from one cell to another cell. When it is going to the new cell it should announce the column headers of the new cell.

### NVDA logs, crash dumps and other attachments:
[[2137516_NVDA is not announcing headers as headers in table in CKEditor demo (1).zip](https://github.com/nvaccess/nvda/files/9504016/2137516_NVDA.is.not.announcing.headers.as.headers.in.table.in.CKEditor.demo.1.zip)](https://microsoftapc-my.sharepoint.com/:u:/g/personal/v-deenthakur_microsoft_com/EaKv3yxacN1MmmAJ7Wui7DEBIrBRE87w_QnNbzrbUP3kGQ?e=AddeMk)

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
NVDA Version 2022.2

#### Windows version:
Windows 11 Version 10.0.22000 Build 22000

#### Name and version of other software in use when reproducing the issue:
Microsoft Edge Version 105.0.1343.27 (Official build) (64-bit)

#### Other information about your system:
N/A

### Other questions
N/A

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

### PR DESCRIPTION
**Pull request by SaschaCowley**
### Link to issue number:

Fixes #14113 

### Summary of the issue:

When navigating tables in contenteditable HTML elements in focus mode, table row and column headers are not reported.

### Description of user facing changes

Table row and column headers are now reported when navigating tables in contenteditable HTML elements in focus mode.

### Description of development approach

Added appropriate fields to the `TextInfos.ControlField` returned by `CompoundTextInfo._getControlFieldForObject`. Also added special case handling for Chromium, which erroneously reports heder cells as being their own headers.

### Testing strategy:

Ran unit tests. Manually tested with a basic contenteditable table with row and column headers in Firefox and Edge.

### Known issues with pull request:

None.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation -- unnecessary
  - Developer / Technical Documentation -- unnecessary
  - Context sensitive help for GUI changes -- unnecessary
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken. -- None should be needed

### CODE REVIEW

**Comment by Adriani90** [comment]
ctrl+alt+arrow keys navigates html tables in browse mode, not in focus mode. I guess the description in the referenced issue is talking about browse mode. Or am I wrong?

**Comment by Adriani90** [comment]
> In Edge, row and column headers are reported even when on the row header or column header cell, meaning the cell's contents are spoken twice.

Is this also happening when using UIA for Edge and Chrome in NVDA advanced settings?

**Comment by SaschaCowley** [comment]
> ctrl+alt+arrow keys navigates html tables in browse mode, not in focus mode. I guess the description in the referenced issue is talking about browse mode. Or am I wrong?

Ctrl+Alt+arrow keys navigates focused tables in focus mode, at least in contenteditable tables. See for example [the demo provided by Mick in the original issue](https://github.com/nvaccess/nvda/files/13700459/test_contenteditable_table.html.txt)

> This should be tested in the Wordweb as well, so using an MS Office 365 account with Word Online should report table headers both in focus and in browse mode ideally. 

Great, I'll give this a try and report back.


> > In Edge, row and column headers are reported even when on the row header or column header cell, meaning the cell's contents are spoken twice.
> 
> Is this also happening when using UIA for Edge and Chrome in NVDA advanced settings?

When forcing the use of UIA in Edge, the double speaking of row and column headers does not occur. However, navigation by Alt+Ctrl+arrow keys stops working. I suspect this is a case of those scripts not being bound in that context. I can't seem to find an issue about this, does one exist already that we know of, or should I go ahead and creat one?

**Comment by michaelDCurran** [comment]
I don't think this PR needs to worry about Edge with UIA -- there are multiple issues there, which should be handled separately. Specifically:
* NVDA does not provide table navigation commands  on editable text or contenteditable nodes in Chromium when using UIA. This could in theory be solved by  having `NVDAObjects.UIA.chromium.ChromiumUIA` also inherit from `documentBase.DocumentWithTableNavigation`. 
* A quick test also shows that  the `_getTableCellAt` method which DocumentWithTableNavigation expects, may not yet be implemented for UIA.
* The same code that was made in this PR (exposing row header and column header text) would also need to be made in `NVDAObjects.UIA.UIATextInfo._getControlFieldForUIAObject`.

However, the issue with multiple reporting of row and column headers on header elements in Edge / chrome (without using UIA) probably should be handled in this PR (or another that blocks this one), as it becomes more noticeable.
Technically I think the bug is with Chromium for inappropriately exposing the same IAccessible2 object as its own column or row header.
But we could work around this perhaps in `NVDAObjects.IAccessible.IAccessible._tableHeaderTextHelper` by filtering out headers that have the same `IAccessible2::uniqueID` as the object whose headers we are fetching.

**Comment by SaschaCowley** [comment]
> However, the issue with multiple reporting of row and column headers on header elements in Edge / chrome (without using UIA) probably should be handled in this PR (or another that blocks this one), as it becomes more noticeable. Technically I think the bug is with Chromium for inappropriately exposing the same IAccessible2 object as its own column or row header. But we could work around this perhaps in `NVDAObjects.IAccessible.IAccessible._tableHeaderTextHelper` by filtering out headers that have the same `IAccessible2::uniqueID` as the object whose headers we are fetching.

I have implemented this, and all seems to be working properly (based on manual and unit testing).

File: source/NVDAObjects/IAccessible/__init__.py
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py:1340]: There is an IA2UniqueID attribute already cached on the IAccessible NVDAObject at construction time. It will be None if not available.
Thus self.IA2UniqueID can be used directly here rather than queryInterface and fetching again.

File: source/NVDAObjects/IAccessible/__init__.py
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py:1359]: Best to keep the minimum amount of actions in a try block so as to not accidentally catch unrelated errors.
Thus Just fetch the uniqueID as headerUniqueID, setting it to None in the except clause. Then do a separate if block after which checks equality and continues if equal.

File: source/NVDAObjects/IAccessible/__init__.py
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py:1362]: You could log a debugWarning message here as although this error is not serious, it is still unexpected.

File: source/NVDAObjects/IAccessible/__init__.py
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py:1364]: It would be good to avoid queryInterface to IAccessible2 twice. Probably best to queryInterface to IAccessible2 on the header much further up, so that fetching of uniqueID and accName can be done on the same one.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
Looking good. But a few review comments.

**Comment by SaschaCowley** [comment]
> This should be tested in the Wordweb as well, so using an MS Office 365 account with Word Online should report table headers both in focus and in browse mode ideally. Word Online is build based on content editable HTML elements, so it is a quite good use case to test content editable HTML elements support in NVDA.
I am getting errors unrelated to this PR (Identical when running from source or 2023.3) when attempting to navigate tables in focus mode in Word Web in Edge, so I suggest this not be a major point of testing for now. The PR seems to be working otherwise.

File: source/NVDAObjects/IAccessible/__init__.py
Comment (michaelDCurran) [inline source/NVDAObjects/IAccessible/__init__.py:1352]: Please catch possible COMError around this line, and if so, log a debugWarning and continue.

**Review by michaelDCurran — DISMISSED** [review]

**Comment by michaelDCurran** [comment]
My understanding is that Word 365 online exposes tables as ARIA live 
regions and not as proper contenteditable structure yet. Office 365 
Online still has a ways to go. Thus, I agree with Sascha that this pr 
should not worry about that.

File: source/NVDAObjects/IAccessible/__init__.py
Comment (gerald-hartig) [inline source/NVDAObjects/IAccessible/__init__.py:2]: Thanks for updating the copyright, you're already covered with the reference to NV Access.

```suggestion
# Copyright (C) 2006-2024 NV Access Limited, Babbage B.V., Cyrille Bougot
```

File: source/compoundDocuments.py
Comment (gerald-hartig) [inline source/compoundDocuments.py:4]:
```suggestion
# Copyright (C) 2010-2024 NV Access Limited, Bram Duvigneau

```

**Review by gerald-hartig — APPROVED** [review]

**Comment by gerald-hartig** [comment]
Although I've approved it, please make sure you get @seanbudd to sign off on it too!

File: user_docs/en/changes.t2t
Comment (michaelDCurran) [inline user_docs/en/changes.t2t:9]:
```suggestion
```

File: user_docs/en/changes.t2t
Comment (michaelDCurran) [inline user_docs/en/changes.t2t:16]:
```suggestion
```

**Review by michaelDCurran — COMMENTED** [review]
Looks like some extra blank lines were added to changes.t2t?

**Comment by seanbudd** [comment]
Please add "Fixes" before the issue number so GitHub automatically links the PR and issue.
Also please mark the final code review checklist items as checked, I assume you have manually tested the latest code changes?

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
This PR can be merged when you are ready and the description is up to date.
When merging PRs we generally perform a "squash merge" and copy paste the PR description into the commit message.
We use the following sections in the commit message:
- Link to issue number
- Summary of the issue
- Description of user facing changes
- Description of development approach
- And if there are known issues, Known issues with pull request

So usually this is just copying the first 4 items, and sometimes the known issues.
You can check the master commit history for examples

**Comment by SaschaCowley** [comment]
> Please add "Fixes" before the issue number so GitHub automatically links the PR and issue.
Where do you mean?

**Comment by seanbudd** [comment]
No problem:

For example:

```
### Link to issue number:
Fixes #14113
```

**Comment by josephsl** [comment]
Hi,

Wait, how was the merge possible (was Sasha approved to merge this PR)? Isn't it usually NV Access folks who can merge PR's? Just curious.
Thanks.

**Comment by seanbudd** [comment]
@SaschaCowley is our newest member of NV Access - I think further introductions will come in future

**Comment by cary-rowen** [comment]
wow, welcome @SaschaCowley

**Comment by SaschaCowley** [comment]
Thanks, @josephsl and @cary-rowen, I'm excited to be here.

**Comment by XLTechie** [comment]
Welcome, @SaschaCowley! A welcome surprise!

**Pull request merged**
